rfd/0166-tpm-joining.md

```diff
@@ -0,0 +1,390 @@
+- [TCG Trusted Platform Module Library](https://trustedcomputinggroup.org/resource/tpm-library-specification/): For specification of TPM commands and structures.
+- [TPM 2.0 Keys for Device identity and Attestation](https://trustedcomputinggroup.org/wp-content/uploads/TPM-2p0-Keys-for-Device-Identity-and-Attestation_v1_r12_pub10082021.pdf): For the recommended TPM 2.0 flows for device identity and attestation.
+- [RFD 0008e - Device Trust TPM Support](https://github.com/gravitational/teleport.e/blob/master/rfd/0008e-device-trust-tpm.md): For the existing TPM identity implementation in Teleport.
+
+Existing code and logic from Device Trust TPM support will be reused where
+appropriate for this feature.
+
+## Overview of TPM functionality and terminology
+
+- Trusted Platform Module (TPM): a hardware module that provides a secure root
+  of trust for a host. It can securely generate and store keys and perform
+  cryptographic operations for the host without exposing the keys.
+- Endorsement Key (EK): A key pair that is "burned in" to the TPM at the time of
+  manufacture. The endorsement key cannot be used to perform signing operations,
+  cannot be exported from the TPM, but can be used to decrypt data.
+- Public Endorsement Key (EKPub): The public part of the EK.
+- Endorsement Key Certificate (EKCert): A certificate containing the EKPub,
+  signed by the TPM manufacturer's CA, that is "burned in" to many TPMs. This 
+  allows the authenticity of the TPM to be verified.
+- Attestation Key (AK): A key pair created for the TPM which can be used to
+  sign data.
+
+A compliant TPM provides certain guarantees that will be relevant to TPM
+joining:
+
+- Certain keys cannot be "exported" from the TPM (e.g available to the host)
+- Certain keys can only be used to perform certain operations (e.g the EK cannot
+  be used to sign or decrypt arbitrary data)
+- The Credential Activation ceremony can be used to prove that a key exists in
+  the TPM with another key, and that the key is configured in a certain way
+  (e.g cannot be exported).
+
+It is important to note that these guarantees do not exist if the TPM is not
+compliant or the host has been compromised in such a way that some malicious
+software is intercepting commands intended for the TPM. This is mitigated by
+validating the EKPub and EKCert against the manufacturer's CA in ceremonies
+such as Credential Activation, proving that the TPM is legitimate and that you 
+are talking directly to it.
+
+## Join Process
+
+### Pre-join Configuration
+
+Before joining is possible, the user must configure the Auth Server with a join
+token which allows that TPM to join. The Join Token will contain the EKPub
+hashes and EKCert serials which are allowed to join.
+
+To assist in this process, a new command will be added `tbot tpm identify`:
```

Comment:
Any process escalation concerns with this command? Will we use the same techniques as device trust? (sudo for Linux, spawned admin terminal for Windows)

Reply:
I'll add a section on this since I haven't been super clear - in this case, sudo isn't strictly necessary. I'd probably instead that the user that `tbot` runs as is granted access to the `tss` group - as it'll need to repeat the Credential Activation for each renewal.

Reply:
Let me retract myself: we use `sudo` to read DMI info from sysfs, but for the actual TPM access we also recommend using the `tss` group. So it may be that you don't need `sudo` at all.

Reply:
Ah ok great - I figured the `tss` route was better.

Reply:
Do note that Windows machines still likely need an admin or escalation. (Not sure how relevant that might be here.)

Reply:
Gotcha, I didn't realize you don't build for Windows. Feel free to resolve this thread.